### ISSUE TITLE
Usage of KL divergence

### ISSUE THREAD

**Issue by Baukebrenninkmeijer** [DAI-Lab/TGAN#24]
I have some questions regarding the paper more than the code specifically. Why did you choose to add the KL divergence as additional loss metric in this way? 

More specifically:
1. Why KL divergence instead of any other (JS, etc.). Am i correct in understanding that your loss essentially consists of `loss = JS + KL_col` where JS represents the original GAN loss which minimizes the JS divergence?
2. If i'm correct, this is done per column. I can see why you would want to model the per column distributions, but shouldn't this already be an result of minimizing the original GAN loss? 
3. Why did you not choose to use a improved already existing loss function like for example Wasserstein?

I've been experimenting with adapting TGAN to use WGAN-GP, but so far, my results have been worse than the original. Variable distributions and column correlations are worse. Given the general gap in GAN and WGAN-GP losses in stability and mode collapse, I was surprised your adaptation works so well. 

Do you have any idea why TGAN would not work well with WGAN-GP, compared to other tasks?

**Comment by leix28** [comment]
Hi, 

(1) There's no specific reason in choosing KL divergence against other loss functions. You can definitely try other distance metrics or feed extra features to the discriminator.
For example, [MedGAN](https://github.com/mp2893/medgan) feeds the mean and standard deviation of a batch as features to the discriminator. [TableGAN](https://github.com/mahmoodm2/tableGAN) explicitly optimizes the each column's mean and standard deviation for generator. 

(2) Not really. Empirically, we found that simply optimize GAN loss will leads to mode collapse. 

(3) Please check our new project, [SDGym](https://github.com/DAI-Lab/SDGym). Our new TGAN model uses WGAN-GP and PacGAN. We apply similar preprocessing method. A paper about SDGym will be available soon. 

Best,
Lei

**Comment by Baukebrenninkmeijer** [comment]
Ah, this is exactly what I've been experimenting with. Using the WGAN-GP architecture. However, in my testing this hasn't improved results, giving worse covariance matrices and worse per column distributions compared to the real dataset. Have you gotten improved results with this? 

The SDGym is exactly something I was thinking of and will be working on for my Thesis, which I'm doing now. I'm very interested in the results of SDGym.
